openssl/src/ssl/mod.rs

```diff
@@ -1434,6 +1438,53 @@ impl<S> SslStream<S> {
         }
     }
 
+    /// Gets the next DTLS handshake timeout.
+    ///
+    /// This is intended to be used with a nonblocking socket, where the value
+    /// can be used to set timeouts on reading and performing handshakes.
+    ///
+    /// `handle_dtlsv1_timeout` should be called when the timeout expires to
+    /// ask OpenSSL to retransmit.
+    #[cfg(any(feature = "dtlsv1", feature = "dtlsv1_2"))]
+    pub fn get_dtlsv1_timeout(&mut self) -> Option<time::Duration> {
+        let mut timeout: timeval = unsafe { mem::zeroed() };
+        let ret = unsafe { ffi::SSL_ctrl(self.ssl.ssl, ffi::DTLS_CTRL_GET_TIMEOUT, 0,
+                                         &mut timeout as *mut _ as *mut _) };
+        if ret > 0 {
+            Some(time::Duration::new(timeout.tv_sec as u64, timeout.tv_usec as u32))
+        } else {
+            None
+        }
+    }
+
+    /// Handles DTLS handshake timeout expiry.
+    ///
+    /// This should be called when the timeout from `get_dtlsv1_timeout`
+    /// expires, to retransmit the handshake.
+    #[cfg(any(feature = "dtlsv1", feature = "dtlsv1_2"))]
+    pub fn handle_dtlsv1_timeout(&mut self) -> Result<(), Error> {
+        let ret = unsafe { ffi::SSL_ctrl(self.ssl.ssl, ffi::DTLS_CTRL_HANDLE_TIMEOUT, 0,
+                                         ptr::null_mut()) };
+        if ret > 0 {
+            Ok(())
+        } else {
+            Err(self.make_error(ret as i32))
+        }
+    }
+
+    /// Performs a TLS/SSL handshake.
+    ///
+    /// This is particularly useful with a nonblocking socket, where the error
+    /// value will identify if OpenSSL is waiting on read or write readiness.
+    pub fn do_handshake(&mut self) -> Result<(), Error> {
```

Comment:
No, but it can be called multiple times for nonblocking sockets – it can return WantRead/WantWrite until calling it again succeeds.